fuzz/delegation/v0.5/fuzzDelegation_test.go

```diff
@@ -0,0 +1,228 @@
+		auctionBalanceAfterUnbond)
+
+	withdrawnAtTheEnd := pfe.getWithdrawTargetBalance()
+
+	totalActiveStake := big.NewInt(0)
+	for delegatorIdx := 0; delegatorIdx <= pfe.numDelegators; delegatorIdx++ {
+		activeUserStake, err := pfe.getUserStakeOfType(delegatorIdx, UserActive)
+		assert.Nil(t, err)
+
+		totalActiveStake = totalActiveStake.Add(totalActiveStake, activeUserStake)
+	}
+
+	activeAndWithdrawn := big.NewInt(0).Add(withdrawnAtTheEnd, totalActiveStake)
+	require.True(t, activeAndWithdrawn.Cmp(pfe.totalStakeAdded) == 0,
+		"Stake added and withdrawn doesn't match. Staked: %d. Withdrawn: %d. Off by: %d",
+		pfe.totalStakeAdded, activeAndWithdrawn,
+		big.NewInt(0).Sub(pfe.totalStakeAdded, activeAndWithdrawn))
+}
+
+func generateRandomEvent(
+	t *testing.T,
+	pfe *fuzzDelegationExecutor,
+	r *rand.Rand,
+	re *fuzzutil.RandomEventProvider,
+	maxDelegationCap *big.Int,
+) {
+	maxStake := big.NewInt(0).Mul(pfe.stakePerNode, big.NewInt(2))
+	maxSystemReward := big.NewInt(1000000000)
+	maxServiceFee := 10000
+	re.Reset()
+
+	switch {
+	case re.WithProbability(0.05):
+		// increment block nonce
+		err := pfe.increaseBlockNonce(r.Intn(1000))
+		require.Nil(t, err)
+	case re.WithProbability(0.05):
+		// add nodes
+		err := pfe.addNodes(r.Intn(3))
+		require.Nil(t, err)
+	case re.WithProbability(0.05):
+		// add nodes
+		err := pfe.removeNodes(r.Intn(2))
+		require.Nil(t, err)
+	case re.WithProbability(0.05):
+		// stake
+		delegatorIdx := r.Intn(pfe.numDelegators + 1)
+		stake := big.NewInt(0).Rand(r, maxStake)
+
+		err := pfe.stake(delegatorIdx, stake)
+		require.Nil(t, err)
+	case re.WithProbability(0.05):
+		ok, err := pfe.isBootstrapMode()
+		require.Nil(t, err)
+
+		if !ok {
+			// add system rewards
+			rewards := big.NewInt(0).Rand(r, maxSystemReward)
+
+			err := pfe.addRewards(rewards)
+			require.Nil(t, err)
+		}
+	case re.WithProbability(0.2):
+		// claim rewards
+		delegatorIdx := r.Intn(pfe.numDelegators + 1)
+
+		err := pfe.claimRewards(delegatorIdx)
+		require.Nil(t, err)
+	case re.WithProbability(0.05):
+		// unStake
+		delegatorIdx := r.Intn(pfe.numDelegators + 1)
+		stake := big.NewInt(0).Rand(r, maxStake)
+
+		err := pfe.unStake(delegatorIdx, stake)
+		require.Nil(t, err)
+	case re.WithProbability(0.05):
+		// unBond
+		delegatorIdx := r.Intn(pfe.numDelegators + 1)
+		err := pfe.unBond(delegatorIdx)
+		require.Nil(t, err)
+	case re.WithProbability(0.05):
+		err := pfe.modifyDelegationCap(big.NewInt(0).Rand(r, maxDelegationCap))
+		require.Nil(t, err)
+
+		err = pfe.continueGlobalOperation()
+		require.Nil(t, err)
+
+		pfe.printServiceFeeAndDelegationCap(t)
+		pfe.printTotalStakeByType()
+	case re.WithProbability(0.05):
+		err := pfe.setServiceFee(r.Intn(maxServiceFee))
+		require.Nil(t, err)
+
+		err = pfe.continueGlobalOperation()
+		require.Nil(t, err)
+
+		pfe.printServiceFeeAndDelegationCap(t)
+		pfe.printTotalStakeByType()
+	case re.WithProbability(0.05):
+		err := pfe.validateOwnerStakeShare()
```

Comment:
validations can be done after each step for extra aggressiveness

Reply:
Done.